Build DD4hep in the CI jobs

msmk0 opened the pull request:
Update CMake config to include DD4hep in the `ACTS_BUILD_EVERYTHING` flag that is used in the CI builds.

Increase the required DD4hep version to 1.10 (available in LCG95apython3).

Fix the CMake config for using Geant4.

Closes #129.

msmk0 commented:
Depends on an updated CI image that is currently being build.

msmk0 commented:
@paulgessinger Could you try to reproduce the build issue locally using the updated container image? I can not reproduce the error in the ubuntu build, when trying to build the object in question (ActsGenParticleGun).

HadrienG2 commented:
@msmk0 I got a very similar error on my system, but for a different example (one of the TGeo ones, I don't remember which). Unfortunately, I have since trashed that build as I was running short on SSD space, so I cannot experiment further...

HadrienG2 commented:
Note, however, that the linker command is pretty weird:

```
: && /usr/bin/c++  -Werror -Wall -Wextra -Wpedantic -Wshadow -Wunused-local-typedefs -O3 -DNDEBUG   Examples/Run/EventGenerator/CMakeFiles/ActsGenParticleGun.dir/ParticleGun.cpp.o  -o bin/ActsGenParticleGun  -Wl,-rpath,/__w/acts/acts/build/lib:/usr/local/lib:/__w/acts/acts/build/Examples/Detectors/DD4hepDetector:  lib/libActsExamplesCommon.so  lib/libActsExamplesGenerators.so  lib/libActsExamplesIoRoot.so  lib/libActsExamplesIoCsv.so  lib/libActsExamplesPrinters.so  /usr/lib/x86_64-linux-gnu/libboost_program_options.so  lib/libActsExamplesTruthTracking.so  lib/libActsExamplesMagneticField.so  lib/libActsExamplesMaterialMapping.so  lib/libActsExamplesGeant4.so  lib/libActsExamplesDD4hepG4.so  /usr/local/lib/libG4Tree.so  /usr/local/lib/libG4GMocren.so  /usr/local/lib/libG4FR.so  /usr/local/lib/libG4visHepRep.so  /usr/local/lib/libG4RayTracer.so  /usr/local/lib/libG4VRML.so  /usr/local/lib/libG4vis_management.so  /usr/local/lib/libG4modeling.so  /usr/local/lib/libG4interfaces.so  /usr/local/lib/libG4persistency.so  /usr/local/lib/libG4error_propagation.so  /usr/local/lib/libG4readout.so  /usr/local/lib/libG4physicslists.so  /usr/local/lib/libG4run.so  /usr/local/lib/libG4event.so  /usr/local/lib/libG4tracking.so  /usr/local/lib/libG4parmodels.so  /usr/local/lib/libG4processes.so  /usr/local/lib/libG4analysis.so  /usr/lib/x86_64-linux-gnu/libz.so  /usr/lib/x86_64-linux-gnu/libexpat.so  /usr/local/lib/libG4digits_hits.so  /usr/local/lib/libG4track.so  /usr/local/lib/libG4particles.so  /usr/local/lib/libG4geometry.so  /usr/local/lib/libG4materials.so  /usr/local/lib/libG4graphics_reps.so  /usr/local/lib/libG4intercoms.so  /usr/local/lib/libG4global.so  /usr/local/lib/libG4clhep.so  Examples/Detectors/DD4hepDetector/libActsExamplesDetectorDD4hep.so  lib/libActsDD4hepPlugin.so  lib/libActsTGeoPlugin.so  /usr/local/lib/libDDCore.so.1.11  /usr/local/lib/libTree.so  /usr/local/lib/libImt.so  /usr/local/lib/libNet.so  /usr/local/lib/libGeom.so  /usr/local/lib/libGenVector.so  /usr/local/lib/libDDParsers.so.1.11  /usr/lib/x86_64-linux-gnu/libxerces-c.so  /usr/local/lib/libRint.so  /usr/local/lib/libPhysics.so  lib/libActsExamplesIoJson.so  lib/libActsJsonPlugin.so  lib/libActsExamplesIoObj.so  -pthread  /usr/lib/x86_64-linux-gnu/libboost_program_options.so  lib/libActsExamplesFramework.so  /usr/local/lib/libHist.so  /usr/local/lib/libMatrix.so  /usr/local/lib/libMathCore.so  /usr/local/lib/libRIO.so  /usr/local/lib/libThread.so  /usr/local/lib/libCore.so  -lpthread  lib/libActsFatras.so  lib/libActsDigitizationPlugin.so  lib/libActsCore.so && :
```

I have no idea why this basic example links against so many things to begin with. But if you take a closer look, you'll notice that `libActsExamplesDD4hepG4.so` gets linked in, as well as all of Geant4 `libActsExamplesDetectorDD4hep.so`, `libDDCore.so.1.11`, `libDDParsers.so.1.11`... but nothing that looks like a DDG4 library. So unless DDG4 symbols are somehow in libDDCore or libDDParsers, it's maybe not surprising that DDG4 symbols are missing, as reported in the build error :

```
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Converter::Geant4Converter(dd4hep::Detector const&, dd4hep::PrintLevel)'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4GeometryInfo::world() const'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Converter::~Geant4Converter()'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Mapping::attach(dd4hep::sim::Geant4GeometryInfo*)'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Mapping::detach()'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Converter::create(dd4hep::DetElement)'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Mapping::instance()'
/usr/bin/ld: lib/libActsExamplesDD4hepG4.so: undefined reference to `dd4hep::sim::Geant4Mapping::volumeManager() const'
```

Of course, the million dollar question is, again, why does ActsGenParticleGun need to link against the entire world like that?

paulgessinger commented:
The link error is most likely DD4hep not populating `${DD4hep_LIBRARIES}` anymore. You can switch that to `DD4hep::DDCore`.

msmk0 commented:
It should not and I thought it does not. Since which version are the DD4hep targets available? LCG95 only provide 1.10.

paulgessinger commented:
See https://gitlab.cern.ch/acts/acts-core/-/merge_requests/798 I guess. In that spot I had to put in a guard to support both cases.

msmk0 commented:
I'm trying around with this right now.

msmk0 commented:
@HadrienG2 @paulgessinger I do not understand the remaining linker errors on the LCG build. Does anyone of you has seen that before or has a hunch as to what is going wrong there?

HadrienG2 commented:
@msmk0 Sorry, never got that one yet!

msmk0 commented:
After some googling this looks to be related to whether Geant4 is compiled w/ or w/o multithreading. There might be some compile flags that are needed in the latter case.

msmk0 commented:
Ready to be merged.

HadrienG2 commented on Examples/Run/MaterialMapping/GeantinoRecording/GeantinoRecordingExample.cpp:
```diff
@@ -16,10 +16,10 @@
-#include "ACTFW/Plugins/DD4hepG4/DD4hepToG4Svc.hpp"
 #include "ACTFW/Utilities/Paths.hpp"
+#include "ActsExamples/Geant4/DD4hepToG4Svc.hpp"
+#include "ActsExamples/Geant4/GeantinoRecording.hpp"
```
Is it right to change some of these paths, but not all of them?

msmk0 replied:
Here, I only changed the paths that I had to move anyways. I am not sure it is right, but you have to start somewhere and I wanted to avoid bloating this PR with unrelated changes. I expect that we will have mix of conventions for a while until everything is converted.

paulgessinger replied:
Yeah I think I agree, I'd have preferred the renaming to not be in this PR.

msmk0 replied:
Ok, I'll switch it back to the original names.

msmk0 replied:
The Geant4 consolidation work has been removed. I'll put that into a separate PR.

HadrienG2 commented on Examples/Algorithms/Geant4/src/GeantinoRecording.cpp:
```diff
@@ -6,23 +6,24 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-#include "ACTFW/MaterialMapping/GeantinoRecording.hpp"
+#include "ActsExamples/Geant4/GeantinoRecording.hpp"
```
Is it right to rename this one, but not the other ACTFW headers?

msmk0 replied:
See above

paulgessinger commented:
I’ll have a look tomorrow.

paulgessinger commented on CMakeLists.txt:
```diff
@@ -103,7 +103,7 @@ find_package(Eigen 3.2.9 REQUIRED)
 # trees to sort of when a particular package is actually needed.
 #
 # version variables for external dependencies that are required more than once
-set(acts_dd4hep_version 1.2)
```
We only required 1.2 before??

msmk0 replied:
I think this was just never updated.

paulgessinger commented on Examples/Algorithms/Geant4/src/MMPrimaryGeneratorAction.cpp:
```diff
@@ -6,18 +6,17 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-#include "ACTFW/Plugins/Geant4/MMPrimaryGeneratorAction.hpp"
+#include "MMPrimaryGeneratorAction.hpp"
```
Why relative includes here? Same for the other cases after this?

msmk0 replied:
Because they are internal files now and not part of the public interface anymore.

msmk0 replied:
Does not apply anymore since the relevant changes have been removed from this PR.

msmk0 commented:
> So no more DD4hepG4 algorithm?

In order to fix the Geant4 linker issues, I moved all Geant4 files into a single module. The `DD4hepG4` module did not contain any algorithms, just some tools, and was only used by the `Geant4` module. The `Geant4` module was also just providing tools that were then used by the MaterialMapping module which had the `GeantinoRecording` algorithm that used only Geant4/DD4hep tools and is actually independent from the rest of the material mapping code.

All three modules had to have some form of Geant4 linkage and each was doing it inconsistently. That is why I moved all Geant4-related code into a single module and hide all internal code, so I could correctly set the Geant4 CMake settings in a single place.

HadrienG2 commented on Examples/Algorithms/Geant4/CMakeLists.txt:
```diff
@@ -9,9 +7,13 @@ add_library(
 target_include_directories(
   ActsExamplesGeant4
-  PUBLIC
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-    ${Geant4_INCLUDE_DIRS})
+  PUBLIC ${Geant4_DEFINITIONS})
```
Shouldn't that be a `target_compile_definitions`?

(Just looking at changes to other CMake files, don't mind me if I'm just showing my CMake ignorance.)

msmk0 replied:
Yes, it should. Thanks for catching that. I has been fixed now.
